packages/back-end/src/services/clickhouse.ts

```diff
@@ -0,0 +1,140 @@
+}
+
+function ensureClickhouseEnvVars() {
+  if (
+    !CLICKHOUSE_HOST ||
+    !CLICKHOUSE_ADMIN_USER ||
+    !CLICKHOUSE_ADMIN_PASSWORD ||
+    !CLICKHOUSE_DATABASE ||
+    !CLICKHOUSE_MAIN_TABLE
+  ) {
+    throw new Error(
+      "Must specify necessary environment variables to interact with clickhouse."
+    );
+  }
+}
+
+export async function createClickhouseUser(
+  context: ReqContext,
+  datasourceId: string
+): Promise<DataSourceParams> {
+  // Temporary protection to prevent users from manually changing feature flag to create Clickhouse users.
+  if (ENVIRONMENT === "production" && context.org.id != "org_24yyifrkf649iz6") {
+    throw new Error(
+      "Clickhouse user creation is only allowed for the Growthbook organization"
+    );
+  }
+
+  ensureClickhouseEnvVars();
+
+  const client = createClickhouseClient({
+    host: CLICKHOUSE_HOST,
+    username: CLICKHOUSE_ADMIN_USER,
+    password: CLICKHOUSE_ADMIN_PASSWORD,
+    database: CLICKHOUSE_DATABASE,
+    application: "GrowthBook",
+    request_timeout: 3620_000,
+    clickhouse_settings: {
+      max_execution_time: 3600,
+    },
+  });
+
+  const orgId = context.org.id;
+  const user = clickhouseUserId(orgId, datasourceId);
+  const password = generator.generate({
+    length: 30,
+    numbers: true,
+  });
+  const hashedPassword = crypto
+    .createHash("sha256")
+    .update(password)
+    .digest("hex");
+
+  const database = user;
+  const viewName = `${database}.events`;
+
+  logger.info(`creating Clickhouse database ${database}`);
+  await client.command({
+    query: `CREATE DATABASE ${database}`,
+  });
+
+  logger.info(`Creating Clickhouse user ${user}`);
+  await client.command({
+    query: `CREATE USER ${user} IDENTIFIED WITH sha256_hash BY '${hashedPassword}'`,
+  });
+
+  logger.info(`Creating Clickhouse view ${viewName}`);
+  await client.command({
+    query: `CREATE VIEW ${viewName} DEFINER=CURRENT_USER SQL SECURITY DEFINER AS SELECT * FROM ${CLICKHOUSE_MAIN_TABLE} WHERE organization = '${orgId}'`,
+  });
+
+  logger.info(`Granting select permissions on ${viewName} to ${user}`);
+  await client.command({ query: `GRANT SELECT ON ${viewName} TO ${user}` });
+
+  logger.info(`Clickhouse user ${user} created`);
+
+  const parts = CLICKHOUSE_HOST.split(":");
+  const port = parseInt(CLICKHOUSE_HOST.split(":").pop() || "9000");
+  const url = parts.join(":");
+
+  const params = {
+    port: port,
+    url: url,
+    user: user,
+    password: password,
+    database: database,
+  };
+
+  return params;
+}
+
+export async function deleteClickhouseUser(
+  datasourceId: string,
+  organization: string
+) {
+  ensureClickhouseEnvVars();
+
+  const client = createClickhouseClient({
+    host: CLICKHOUSE_HOST,
+    username: CLICKHOUSE_ADMIN_USER,
+    password: CLICKHOUSE_ADMIN_PASSWORD,
+    database: CLICKHOUSE_DATABASE,
+    application: "GrowthBook",
+    request_timeout: 3620_000,
+    clickhouse_settings: {
+      max_execution_time: 3600,
+    },
+  });
```

Comment:
I would abstract out creating the admin clickhouse client into a separate function since it's shared in 2 places (and likely more in the future).